tests/federation_room_upgrade_test.go

```diff
@@ -0,0 +1,123 @@
+package tests
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+// NOTE: this test falls under a "should" guarantee from the spec: https://spec.matrix.org/v1.4/client-server-api/#server-behaviour-16
+// > When a user joins the new room, the server SHOULD automatically transfer/replicate some of the user’s personalized settings such as notifications, tags, etc.
+//
+// sytest: $user_type user has push rules copied to upgraded room
+func TestRoomUpgradeCopiesPushRules(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintFederationTwoLocalOneRemote)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+	charlie := deployment.Client(t, "hs2", "@charlie:hs2")
+
+	doTestWith := func(t *testing.T, user *client.CSAPI) {
+		roomID := alice.CreateRoom(t, map[string]interface{}{
+			"preset": "public_chat",
+		})
+
+		user.JoinRoom(t, roomID, []string{"hs1"})
+		alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(user.UserID, roomID))
+
+		user.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "pushrules", "global", "room", roomID},
+			client.WithJSONBody(t, map[string]interface{}{
+				"actions": []string{"notify"},
+			}),
+		)
+
+		_, sinceToken := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
```

Comment:
Why is alice syncing here? Please make the condition explicit

Reply:
This is because alice upgrades the room, and will be the first to see the new events and its results. IIRC this is also what the sytest did, incrementally syncing from a room's upgrader to observe its results there.

Reply:
My objection isn't that the user that's syncing is Alice; it's that I don't know what we are waiting for Alice to see down /sync before upgrading the room. That's why I suggest splitting out `user` and `alice`'s syncs in https://github.com/matrix-org/complement/pull/554/files#r1047211314 --- it makes it easier to see what each client is waiting to see before the scenario progresses.